src/marqo/core/inference/device_manager.py

```diff
@@ -0,0 +1,104 @@
+                     f'{self.best_available_device_type.value}.')
+
+    @property
+    def cuda_devices(self):
+        return [device for device in self.devices if device.type == DeviceType.cuda]
+
+    def cuda_device_health_check(self) -> None:
+        """
+        Checks the status of the CUDA devices, and raises exceptions if it becomes
+        not available or out of memory.
+
+        raises
+          - CudaDeviceNotAvailableError if CUDA device is not available.
+          - CudaOutOfMemoryError if any CUDA device is out of memory.
+        """
+        if not self._is_cuda_available_at_startup:
+            # If the instance is initialised without cuda devices, skip the check
+            return
+
+        if not torch.cuda.is_available():
+            # CUDA devices could become unavailable/unreachable if the docker container running Marqo loses access
+            # to the device symlinks. There is no way to recover from this, we will need to restart the container.
+            # See https://github.com/NVIDIA/nvidia-container-toolkit/issues/48 for more details.
+            logger.error('CUDA device/s have become unavailable')
+            raise CudaDeviceNotAvailableError('CUDA device/s have become unavailable')
+
+        oom_errors = []
+        for device in self.cuda_devices:
+            try:
+                cuda_device = torch.device(device.name)
+                memory_stats = torch.cuda.memory_stats(cuda_device)
+                logger.debug(f'Cuda device {device.name} with total memory {device.total_memory}. '
+                             f'Memory stats: {str(memory_stats)}')
+
+                torch.randn(3, device=cuda_device)
+            except RuntimeError as e:
+                if 'out of memory' in str(e).lower():
+                    # `~torch.cuda.empty_cache` doesn't increase the amount of GPU memory available for PyTorch.
+                    # However, it may help reduce fragmentation of GPU memory in certain cases.
+                    torch.cuda.empty_cache()
+
+                    logger.error(f'Cuda device {device.name} is out of memory. Total memory: {device.total_memory}. '
+                                 f'Memory stats: {str(memory_stats)}')
+                    allocated_mem = memory_stats.get("allocated.all.current", None) if memory_stats else None
+                    oom_errors.append(f'Cuda device {device.name} is out of memory:'
+                                      f' ({allocated_mem}/{device.total_memory})')
+                else:
+                    # Log out a warning message when encounter other transient errors.
+                    logger.warning(f'Encountered issue inspecting Cuda device {device.name}: {str(e)}')
+            except Exception as e:
+                # Log out a warning message when encounter other transient errors.
+                logger.warning(f'Encountered issue inspecting Cuda device {device.name}: {str(e)}')
+
+        if oom_errors:
```

Comment:
This only works if all the errors are oom errors. What about the case where we have multiple CUDA devices:
1. gpu 0 is OOM
2. gpu 1 is unavailable/unreachable for some reason (`CudaDeviceNotAvailableError`)
3. gpu 2 has some unknown exception when being inspected (this will go to the `Exception` catch)
4. gpu 3 is fine

Some questions:
- Should we always return a 500 as long as even 1 device has an issue? If this is the case, the error should have the status of each gpu with a problem and clearly state the reason. eg: "device 0 is out of memory (14.9g/15g); device 1 is unavailable; device 2 encountered an error"

- Should we return a 200 as long as there is even 1 healthy device instead?
cc @farshidz

Reply:
> Should we always return a 500 as long as even 1 device has an issue?
Yes, untill we improved Marqo to quarantine the problematic devices.

> Should we return a 200 as long as there is even 1 healthy device instead?
No this will cause Marqo to behaviour weirdly. Some requests will fail but some will succeed. We would rather restart Marqo and make sure all devices are healthy.

Reply:
Makes sense. If this is the case, other errors should be attended to the "error list", not just OOM errors. So when the 500 is returned, it has all issues related to all cuda devices

Reply:
I only error out for OOM issues but not other 'unknown issues' because we don't know if other issues are transient or unrecoverable. I just log them out so we can triage later.